R/sgdnet.R

```diff
@@ -187,10 +190,12 @@ sgdnet.default <- function(x,
                                       "multinomial",
                                       "mgaussian"),
                            alpha = 1,
+                           non_convexity = 1.0,
                            nlambda = 100,
                            lambda.min.ratio =
                              if (NROW(x) < NCOL(x)) 0.01 else 0.0001,
                            lambda = NULL,
+                           penalty = c("", "MCP","SCAD"),
```

Comment:
Hi mentor, I want to make the convex penalty part's default setting the same with glmnet and let previous examples and test cases' input setting stay the same. I think if the non-convexity parameter is not given by user, user need not to rule out MCP or SCAD again, in that case, they will choose lasso, ridge or elastic net by setting alpha.

Reply:
I don't quite understand what you're saying here. 

Isn't it enough to have 

```
penalty = c("lasso", "MCP", "SCAD")
nonconvexity = ifelse(penalty == "MCP", 3, 7)
```

as function arguments and then pass `nonconvexity` in all cases, but just ignore it when applying the lasso prox (soft-thresholding)? 

All of these should devolve to ridge when the appropriate L2 mixing is set to 100%. 

Somewhat relatedly: the mixed case is a bit obnoxious. "LASSO" might be better.

Reply:
Hi mentor, I am sorry that I still have some questions. I mean I want to keep it be consistent with previous setting (without specifying penalty name and use alpha to make choices in elastic net, like glmnet and our vignettes and test cases did) and this option only enables mcp + ridge or scad +ridge with alpha to adjust ridge’s strength. I think I should raise warning if user input MCP/SCAD and set alpha to 0?

Reply:
I'm a bit confused:by "specifying penalty name", are you perhaps not aware of the `match.arg` function in `R`?

If you set `penalty = c("LASSO", "MCP", "SCAD")` in the function definition and then

`penalty <- match.arg(penalty)`

in the function body, if the user doesn't supply a penalty name, it will be `"LASSO"` by default. (`match.arg` also does partial matching, so users can just say `"L"` instead of `"LASSO"`.)

I'm not sure what you mean about elastic net and ridge. If the user sets `alpha = 0`,all three methods should give the same (ridge) answer. Ideally, the code for all three methods should work the same in this case, but if we have a specialized path for ridge, checking `alpha = 0` should activate it regardless of the penalty. 

This doesn't seem that complicated, but we're definitely talking past each other, so if you reply, can you give a super detailed explanation (including links to current code) of what you're worried about.

Reply:
Hi mentor, I am very grateful for your explanations! I am sorry that I totally misunderstood match.arg’s usage, I previously thought empty input was not allowed. I am agree with this setting: ```penalty = c("LASSO", "MCP", "SCAD”)```. And set ```non_convexity``` to 1.0 if no input penalty. For ``alpha = 0``, I think there are some reasons for ncvreg not allowing ```alpha = 0```, so in this case, penalty is set to "LASSO" and non_convexity to 1.0 to use our ridge update, since it is also faster.

Reply:
I'm not sure why it makes a difference in terms of speed, but yeah, that's probably harmless. 

I don't think you need to change `non_convexity` when `penalty == "LASSO"` since it should be ignored anyways.

Reply:
Hi mentor, I set wscale to 1 if non_convexity > 1, if there was a bad input where both “LASSO” and non_convexity > 1, ridge will not be fit, so I want to prevent it.

Reply:
You're talking about this line: https://github.com/jolars/sgdnet/pull/26/files#diff-252fc4768f90a917d4636c72d60d65e5R135 ? 

I don't quite understand what it does, but in general, I would expect the nonconvexity parameter to be fully ignored in a lasso case, so it does look a bit funny.